File: copy-list-with-random-pointer.java
// Time Complexity :
O(3 * length of linked list ) ==> asmptotically o(length of linked list) where m and n are lengths of the two arrays
// Space Complexity :
o(1)
// Did this code successfully run on Leetcode :
yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach


/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
          Node p = head; 
          Node front = head;

          // First round: make copy of each node,
          // and link them together side-by-side in a single list.
          while (p != null) {
            front = p.next;

            Node copy = new Node(p.val);
            p.next = copy;
            copy.next = front;

            p = front;
          }

          // Second round: assign random pointers for the copy nodes.
          p = head;
          while (p != null) {
            if (p.random != null) {
              p.next.random = p.random.next;
            }
            p = p.next.next;
          }

          // Third round: restore the original list, and extract the copy list.
          p = head;
          Node prev = new Node(0);
          Node copy = prev;

          while (p != null) {
            front = p.next.next;

            // extract the copy
            copy.next = p.next;
            copy = copy.next;

            // restore the original list
            p.next = front;

            p = front;
          }

          return prev.next;  
    }
}

File: Copy_List_with_Random_Pointer.py
# Time Complexity : O(N), where N is the number of nodes in the linked list
# Space Complexity : O(1)

class Node:
    def __init__(self, val=0, next=None, random=None):
        self.val = val
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None

        # Step 1: Create a copy of each node and insert it after the original node
        curr = head
        while curr:
            new_node = Node(curr.val, None, None)
            new_node.next = curr.next
            curr.next = new_node
            curr = new_node.next

        # Step 2: Assign random pointers for the copied nodes
        curr = head
        while curr:
            if curr.random:
                curr.next.random = curr.random.next
            curr = curr.next.next

        # Step 3: Separate the original and copied lists
        curr = head
        new_head = head.next
        new_curr = new_head

        while curr:
            curr.next = curr.next.next
            if new_curr.next:
                new_curr.next = new_curr.next.next
            curr = curr.next
            new_curr = new_curr.next

        return new_head